include/libpmemobj++/experimental/concurrent_hash_map.hpp

```diff
@@ -0,0 +1,3063 @@
+#include <libpmemobj++/make_persistent_array.hpp>
+#include <libpmemobj++/make_persistent_array_atomic.hpp>
+#include <libpmemobj++/make_persistent_atomic.hpp>
+#include <libpmemobj++/mutex.hpp>
+#include <libpmemobj++/p.hpp>
+#include <libpmemobj++/persistent_ptr.hpp>
+#include <libpmemobj++/transaction.hpp>
+#if LIBPMEMOBJ_CPP_VG_HELGRIND_ENABLED
+#include <libpmemobj++/detail/common.hpp>
+#endif
+
+#ifndef LIBPMEMOBJ_CPP_USE_TBB_RW_MUTEX
+#define LIBPMEMOBJ_CPP_USE_TBB_RW_MUTEX 0
+#endif
+
+#if LIBPMEMOBJ_CPP_USE_TBB_RW_MUTEX
+#include "tbb/spin_rw_mutex.h"
+#else
+#include <libpmemobj++/shared_mutex.hpp>
+#endif
+
+#include "persistent_pool_ptr.hpp"
+
+#include <atomic>
+#include <cassert>
+#include <functional>
+#include <initializer_list>
+#include <iterator> // for std::distance
+#include <memory>
+#include <mutex>
+#include <thread>
+#include <type_traits>
+
+#if _MSC_VER
+#include <intrin.h>
+#include <windows.h>
+#endif
+
+#ifndef LIBPMEMOBJ_CPP_USE_ATOMIC_ALLOCATOR
+#define LIBPMEMOBJ_CPP_USE_ATOMIC_ALLOCATOR 0
+#endif
+
+namespace std
+{
+/**
+ * Specialization of std::hash for p<T>
+ */
+template <typename T>
+struct hash<pmem::obj::p<T>> {
+	size_t
+	operator()(const pmem::obj::p<T> &x) const
+	{
+		return hash<T>()(x.get_ro());
+	}
+};
+} // namespace std
+
+namespace pmem
+{
+namespace obj
+{
+namespace experimental
+{
+
+using namespace pmem::obj;
+
+/**
+ * hash_compare that is default argument for concurrent_hash_map
+ */
+template <typename Key>
+struct hash_compare {
+	static size_t
+	hash(const Key &a)
+	{
+		return std::hash<Key>{}(a);
+	}
+
+	static bool
+	equal(const Key &a, const Key &b)
+	{
+		return std::equal_to<Key>{}(a, b);
+	}
+};
+
+template <typename Key, typename T, typename HashCompare = hash_compare<Key>>
+class concurrent_hash_map;
+
+/** @cond INTERNAL */
+namespace internal
+{
+template <typename Mutex>
+void
+assert_not_locked(Mutex &mtx)
+{
+#ifndef NDEBUG
+	assert(mtx.try_lock());
+	mtx.unlock();
+#else
+	(void)mtx;
+#endif
+}
+
+template <typename Mutex>
+void
+assert_not_locked(pmem::obj::experimental::v<Mutex> &mtx)
+{
+	assert_not_locked<Mutex>(mtx.get());
+}
+
+template <typename T>
+inline T
+load_with_acquire(const volatile T &location)
+{
+#if LIBPMEMOBJ_CPP_VG_HELGRIND_ENABLED
+	VALGRIND_HG_DISABLE_CHECKING(&location, sizeof(location));
+#endif
+	T ret = location;
+#if LIBPMEMOBJ_CPP_VG_HELGRIND_ENABLED
+	VALGRIND_HG_ENABLE_CHECKING(&location, sizeof(location));
+#endif
+	std::atomic_thread_fence(std::memory_order_acquire);
+
+	return ret;
+}
+
+template <typename T>
+inline void
+store_with_release(volatile T &location, T value)
+{
+#if LIBPMEMOBJ_CPP_VG_HELGRIND_ENABLED
+	VALGRIND_HG_DISABLE_CHECKING(&location, sizeof(location));
+#endif
+	std::atomic_thread_fence(std::memory_order_release);
+	location = value;
+#if LIBPMEMOBJ_CPP_VG_HELGRIND_ENABLED
+	VALGRIND_HG_ENABLE_CHECKING(&location, sizeof(location));
+#endif
+}
+
+#if _MSC_VER
+static inline unsigned long
```

Comment:
this variant of function returns **unsigned long**, where others returns **int**